Don't use progressbars for pull if client is no-tty

LK4D4 opened the pull request:
Fix #3694 
It's only for design review, if it'll be ok, I'll do push as well.

calavera commented:
I think this is a good idea. I'd like to know what @aaronlehmann and @tonistiigi think about it, I know they have some thoughts about the pull command.

tonistiigi commented:
I'm ok with this atm but in the future the pull code should only handle progress messages with channels that transfer objects. So if this work is done this `isTerminal` switch would move to the api side.

aaronlehmann commented:
I'm okay with this as well. As an enhancement, it would be nice to still output progress in a less awkward way if stdout is not a terminal. I'm working on an overhaul of the pull/push code that uses channels to communicate abstract progress information instead of writing JSON directly, which will make this a lot easier.

LK4D4 commented:
@aaronlehmann will that overhaul happen before 1.10?

aaronlehmann commented:
Yeah, I plan to open the PR soon

calavera commented:
@LK4D4, @aaronlehmann should we close this PR now that #18353 is open? Does that one address this issue?

LK4D4 commented:
@calavera no, it's not, I'm waiting for it to update this.

LK4D4 committed:
Don't use progressbars for pull if client is no-tty

Signed-off-by: Alexander Morozov <lk4d4@docker.com>

LK4D4 commented:
@calavera @aaronlehmann I updated PR.
However I see something weird without tty:
```
Using default tag: latest
latest: Pulling from library/busybox
d7e8ec85c5ab: Pulling fs layer
a3ed95caeb02: Pulling fs layer
a3ed95caeb02: Download complete
d7e8ec85c5ab: Download complete
d7e8ec85c5ab: Pull complete
d7e8ec85c5ab: Pull complete
a3ed95caeb02: Pull complete
a3ed95caeb02: Pull complete
Digest: sha256:e4f93f6ed15a0cdd342f5aae387886fba0ab98af0a102da6276eaf24d6e6ade0
Status: Downloaded newer image for busybox:latest
```
You can see that "Pull complete" appear twice for each layer.

aaronlehmann commented:
Hmm. It's a side effect of my transfer manager PR. It does progress notification in an async way so that it doesn't block the actual transfers. This means some updates might be missed. When a transfer finishes, it replays the final update in case it was missed earlier.

Obviously, this was designed with the tty in mind, and it leads to weird output without a tty.

if you want, I can try to revisit this.

aaronlehmann commented on daemon/daemon.go:
```diff
@@ -1122,7 +1122,7 @@ func (daemon *Daemon) PushImage(ref reference.Named, metaHeaders map[string][]st
 
 	go func() {
-		writeDistributionProgress(cancelFunc, outStream, progressChan)
+		writeDistributionProgress(cancelFunc, outStream, progressChan, true)
```
Why not do this for push as well?

Commit:
Avoid outputting last progress item twice

A watcher would output the current progress item when it was detached,
in case it missed that item earlier, which would leave the user seeing
some intermediate step of the operation. This commit changes it to only
output it on detach if it didn't already output the same item.

Signed-off-by: Aaron Lehmann <aaron.lehmann@docker.com>

aaronlehmann commented:
@LK4D4 Does https://github.com/aaronlehmann/docker/commit/417bb5caf19ae0d9e13a7379b10bbb845aa16a87 solve the problem? If so, feel free to include it in this PR.

LK4D4 commented:
@aaronlehmann Will try, thanks!

LK4D4 commented:
@aaronlehmann yup, it fixed it, but now I also see one `Verifying checksum` :)

aaronlehmann commented:
I think that is expected.

LK4D4 commented:
@aaronlehmann Cool. Are you ok with feature overall?

LK4D4 commented:
However, I think this PR is unnecessary, it works in master.
